src/components/Tokens/TokenTable/TokenTable.tsx

```diff
@@ -145,64 +53,93 @@ function NoTokensState({ message }: { message: ReactNode }) {
-  if (topTokens === null) {
-    return (
-      <NoTokensState
-        message={
-          <>
-            <AlertTriangle size={16} />
-            <Trans>An error occured loading tokens. Please try again.</Trans>
-          </>
+  // TODO: consider moving prefetched call into app.tsx and passing it here, use a preloaded call & updated on interval every 60s
+  const prefetchedTokens = usePrefetchTopTokens()
+  const { loading, tokens, loadMoreTokens } = useTopTokens(prefetchedTokens)
+  console.log('loading', loading)
+  const hasMore = !tokens || tokens.length < MAX_TOKENS_TO_LOAD
+
+  const observer = useRef<IntersectionObserver>()
+  const lastTokenRef = useCallback(
+    (node: any) => {
+      if (loading) return
+      if (observer.current) observer.current.disconnect()
+      observer.current = new IntersectionObserver((entries) => {
+        if (entries[0].isIntersecting && hasMore) {
+          loadMoreTokens()
         }
-      />
-    )
-  }
-
-  if (showFavorites && sortedFilteredTokens?.length === 0) {
-    return <NoTokensState message={<Trans>You have no favorited tokens</Trans>} />
-  }
+      })
+      if (node) observer.current.observe(node)
+    },
+    [loading, hasMore, loadMoreTokens]
+  )
 
-  if (!showFavorites && sortedFilteredTokens?.length === 0) {
-    return <NoTokensState message={<Trans>No tokens found</Trans>} />
+  /* loading and error state */
+  if (loading && (!tokens || tokens?.length === 0)) {
+    return <LoadingTokenTable />
+  } else {
+    if (!tokens) {
+      return (
+        <NoTokensState
+          message={
+            <>
+              <AlertTriangle size={16} />
+              <Trans>An error occured loading tokens. Please try again.</Trans>
+            </>
+          }
+        />
+      )
+    } else if (tokens?.length === 0) {
+      return showFavorites ? (
+        <NoTokensState message={<Trans>You have no favorited tokens</Trans>} />
+      ) : (
+        <NoTokensState message={<Trans>No tokens found</Trans>} />
+      )
+    } else {
+      return (
+        <>
+          <GridContainer>
+            <HeaderRow />
+            <TokenDataContainer>
+              {tokens.map((token, index) => {
+                if (tokens.length === index + 1) {
+                  return (
+                    <LoadedRow
+                      key={token?.name}
+                      tokenListIndex={index}
+                      tokenListLength={tokens?.length ?? 0}
+                      token={token}
+                      ref={lastTokenRef}
+                    />
+                  )
+                } else {
+                  return (
+                    <LoadedRow
+                      key={token?.name}
+                      tokenListIndex={index}
+                      tokenListLength={tokens?.length ?? 0}
+                      token={token}
+                    />
+                  )
+                }
+              })}
+              {loading && Array(5).fill(<LoadingRow />)}
```

Comment:
yes it will haha, bad code written to get somethign up n running quickly. new constant defined outside component should fix. 

I chose 5 because this figma file shows 5: https://www.figma.com/file/kNSDBMpOzxSTOP6MerohLm/Web-App-Spec-(Phase-0-%26-Phase-1)-%5BWIP%5D?node-id=2918%3A60584

however we can change it to whatever you think is appropriate?